### PR TITLE
DM-27253: reimplement commonSkyPix overlaps with more general tables

### PR DESCRIPTION
**Pull request by TallJimbo**
This doesn't change the functionality available at a high level; it just defines tables and starts to define interfaces that we can use to materialize more spatial joins in the future.

### CODE REVIEW

File: python/lsst/daf/butler/registry/dimensions/governor.py [hunk 1]
```diff
@@ -64,12 +64,13 @@ def __init__(self, db: Database, dimension: GovernorDimension, table: sqlalchemy
         self._dimension = dimension
```
Comment (andy-slac) [inline python/lsst/daf/butler/registry/dimensions/governor.py:62]: From old commit - lines 50-53 above - duplicate `Parameters` heading.

File: python/lsst/daf/butler/registry/dimensions/overlaps.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+    New instances should be constructed by calling `initialize`, not by calling
+    the constructor directly.
+
+    Parameters
+    ----------
+    db : `Database`
+        Interface to the underlying database engine and namespace.
+    elementStorage : `tuple` [ `DatabaseDimensionRecordStorage` ]
+        Storage objects for the elements this object will related.
```
Comment (andy-slac) [inline python/lsst/daf/butler/registry/dimensions/overlaps.py]: `will related` - `will be related to`? And next parameter too.

File: python/lsst/daf/butler/registry/dimensions/static.py [hunk 3]
```diff
@@ -56,27 +66,74 @@ class StaticDimensionRecordStorageManager(DimensionRecordStorageManager):
...
+        self._overlaps = overlaps
 
     @classmethod
     def initialize(cls, db: Database, context: StaticTablesContext, *,
                    universe: DimensionUniverse) -> DimensionRecordStorageManager:
         # Docstring inherited from DimensionRecordStorageManager.
+        # Start by initializing governor dimensions; those go both in the main
+        # 'records' mapping we'll pass to init, and a local dictionary that we
+        # can pass in when initializing storage for DatabaseDimensionElements.
+        governors: NamedKeyDict[GovernorDimension, GovernorDimensionRecordStorage] = NamedKeyDict()
```
Comment (andy-slac) [inline python/lsst/daf/butler/registry/dimensions/static.py]: Just curious - could there be syntax like:

    governors = NamedKeyDict[GovernorDimension, GovernorDimensionRecordStorage]()

to reduce strain on my eyes?

File: python/lsst/daf/butler/registry/dimensions/table.py [hunk 4]
```diff
@@ -158,13 +228,560 @@ def sync(self, record: DimensionRecord) -> bool:
...
+                levels.sort(reverse=True)
+                # Start with the first level, which is the finest-grained one.
+                # Compute skypix envelope indices directly for that.
+                indices: Dict[int, Set[int]] = {levels[0]: set()}
+                for begin, end in system[levels[0]].pixelization.envelope(record.region):  # type: ignore
+                    indices[levels[0]].update(range(begin, end))
+                # Divide those indices by powers of 4 (and remove duplicates)
+                # work our way up to the last (coarsest) level.
+                for lastLevel, nextLevel in zip(levels[:-1], levels[1:]):
+                    factor = 4**(lastLevel - nextLevel)
```
Comment (andy-slac) [inline python/lsst/daf/butler/registry/dimensions/table.py:693]: Is this true for all known pixelization systems? Or do we only care about HTM?

**Review by andy-slac — APPROVED** [review]
Looks OK, though it's a bit dense in places and I'm not sure I could understand every line of it 🙂

File: python/lsst/daf/butler/registry/dimensions/overlaps.py (hunk 2 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/registry/dimensions/overlaps.py]: :+1: was supposed to be "will relate".

File: python/lsst/daf/butler/registry/dimensions/static.py (hunk 3 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/registry/dimensions/static.py]: Good point!  That syntax doesn't always work (for example with `Optional` or `Union`), so I always forget to use it when it does.  I will here.

File: python/lsst/daf/butler/registry/dimensions/table.py (hunk 4 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/registry/dimensions/table.py:693]: I am fairly certain it is true for HTM, Q3C, and healpix (for the "nested" healpix numbering system that we'd use, at least).

**Pull request merged**